llvm/lib/Target/RISCV/RISCVInstrInfo.cpp

```diff
@@ -1619,8 +1619,223 @@ static bool isFMUL(unsigned Opc) {
+  case RISCV::OPC##_M4_MASK:                                                   \
+  case RISCV::OPC##_M8_MASK:                                                   \
+  case RISCV::OPC##_MF2_MASK:                                                  \
+  case RISCV::OPC##_MF4_MASK:                                                  \
+  case RISCV::OPC##_MF8_MASK
+
+  unsigned Opcode = Inst.getOpcode();
+  if (Invert) {
+    if (auto InvOpcode = getInverseOpcode(Opcode))
+      Opcode = *InvOpcode;
+    else
+      return false;
+  }
+
+  // clang-format off
+  switch (Opcode) {
+  default:
+    return false;
+  OPCODE_LMUL_CASE(PseudoVADD_VV):
+  OPCODE_LMUL_MASK_CASE(PseudoVADD_VV):
+  OPCODE_LMUL_CASE(PseudoVMUL_VV):
+  OPCODE_LMUL_MASK_CASE(PseudoVMUL_VV):
+    return true;
+  }
+  // clang-format on
+
+#undef OPCODE_LMUL_MASK_CASE
+#undef OPCODE_LMUL_CASE
+}
+
+bool RISCVInstrInfo::areRVVInstsReassociable(const MachineInstr &MI1,
+                                             const MachineInstr &MI2) const {
+  if (!areOpcodesEqualOrInverse(MI1.getOpcode(), MI2.getOpcode()))
+    return false;
+
+  assert(MI1.getMF() == MI2.getMF());
+  const MachineRegisterInfo *MRI = &MI1.getMF()->getRegInfo();
+  const TargetRegisterInfo *TRI = MRI->getTargetRegisterInfo();
+
+  // Make sure vtype operands are also the same.
+  const MCInstrDesc &Desc = get(MI1.getOpcode());
+  const uint64_t TSFlags = Desc.TSFlags;
+
+  auto checkImmOperand = [&](unsigned OpIdx) {
+    return MI1.getOperand(OpIdx).getImm() == MI2.getOperand(OpIdx).getImm();
+  };
+
+  auto checkRegOperand = [&](unsigned OpIdx) {
+    return MI1.getOperand(OpIdx).getReg() == MI2.getOperand(OpIdx).getReg();
+  };
+
+  // PassThru
+  if (!checkRegOperand(1))
+    return false;
+
+  // SEW
+  if (RISCVII::hasSEWOp(TSFlags) &&
+      !checkImmOperand(RISCVII::getSEWOpNum(Desc)))
+    return false;
+
+  // Mask
+  if (RISCVII::usesMaskPolicy(TSFlags)) {
+    const MachineBasicBlock *MBB = MI1.getParent();
+    const MachineBasicBlock::const_reverse_iterator It1(&MI1);
+    const MachineBasicBlock::const_reverse_iterator It2(&MI2);
+    Register MI1VReg;
+
+    bool SeenMI2 = false;
+    for (auto End = MBB->rend(), It = It1; It != End; ++It) {
+      if (It == It2) {
+        SeenMI2 = true;
+        if (!MI1VReg.isValid())
+          // There is no V0 def between MI1 and MI2; they're sharing the
+          // same V0.
+          break;
+      }
+
+      if (It->definesRegister(RISCV::V0, TRI)) {
+        Register SrcReg =
+            TRI->lookThruCopyLike(It->getOperand(1).getReg(), MRI);
+
+        if (!MI1VReg.isValid()) {
+          // This is the V0 def for MI1.
+          MI1VReg = SrcReg;
+          continue;
+        }
+
+        // Some random mask updates.
+        if (!SeenMI2)
+          continue;
+
+        // This is the V0 def for MI2; check if it's the same as that of
+        // MI1.
+        if (MI1VReg != SrcReg)
```

Comment:
I think the SrcReg should always be of a virtual register, but should we assert it here just in case?

Reply:
I don't think SrcReg will necessarily be a virtual register. For example, if the mask is from a function argument, something like this will be generated:
```
    %1:vr = COPY $v0
    $v0 = COPY %1
    %3:vrnov0 = PseudoVSRL_VI_MF8_MASK $noreg, %0, 1, $v0, %2, 3 /* e8 */, 1 /* ta, mu */
    $v0 = COPY %1
    %5:vrnov0 = PseudoVAND_VX_MF8_MASK $noreg, killed %3, killed %4, $v0, %2, 3 /* e8 */, 1 /* ta, mu */
   ...
```
And SrcReg will be V0 in this case.

Reply:
In that case don't we need to check that it's a virtual register?

Reply:
> In that case don't we need to check that it's a virtual register?

Using my snippet in the previous comment as an example, assuming `%3` is MI1 and `%5` is MI2, SrcReg will be the result of `TRI->lookThruCopyLike(%1, ...)` in which `%1` comes from the operand of either of the later two COPYs.  And since we look through COPY, SrcReg will be `$v0` from the first line of the snippet. So I don't think SrcReg will necessary be virtual register (unless you're not talking about SrcReg but suggesting to check another register)

Reply:
Sorry, I agree that it might not be a virtual register, but what I meant to say was that if SrcReg is a physical register then we would need to check that nothing writes to it in between M1 and M2.

Reply:
> we would need to check that nothing writes to it in between M1 and M2.

Absolutely, and I think I've done that: the loop starting from 1704 is traversing backward from M1, trying to find the last V0 def before M1. It continues and when traversing pass M2, it tried to find the last V0 def before M2, comparing it with that of before M1 which we have retrieved. 
For cases where there is no V0 def between M1 and M2, the check on line 1707 covers that.

Reply:
Now I think of it, my logics won't cover cases like:
```
// M2
%2 = .... // modify mask
$v0 = COPY %2
%3 = COPY $v0
$v0 = COPY %3
// M1
```
Though does anyone even generate code like this? My understanding is that if you change mask between M1 and M2, it should result in something like:
```
// M2
%2 = .... // modify mask
$v0 = COPY %2
// M1
```
Which my logics is able to handle correctly.

Reply:
My concern was that SrcReg might be a physical register that's not V0, hopefully that doesn't happen in practice but if it did then we're not checking for any other physical register defs. e.g.

```
$v1 = FOO
$v0 = COPY $v1
M2 $v0
$v1 = BAR
$v0 = COPY $v1
M1 $v0
```

I think SelectionDAG will always generate a copy from a virtual register to V0 before each masked instruction though, i.e. it should look like:

```
$v0 = COPY %foo
M2 $v0
$v0 = COPY %bar
M1 $v0
```

So we can probably get away with just using the copy's source without `lookThruCopyLike`, making sure to check that it's a virtual register. I think that should handle the function argument example you mentioned above.

Reply:
> just using the copy's source without lookThruCopyLike, making sure to check that it's a virtual register.

yeah I think this is a good middle ground. I've just updated the PR with this solution.